src/relax/transform/lift_transform_params.cc

```diff
@@ -494,22 +700,71 @@ class ConsumeBundledParams : public ExprMutator {
   std::unordered_map<int, Expr> param_remap_;
 };
 
+std::vector<std::pair<GlobalVar, Function>> GetTargetFunctions(
+    const IRModule& mod, const Array<String>& target_function_names) {
+  std::vector<std::pair<GlobalVar, Function>> target_functions;
+  if (target_function_names.size()) {
+    for (const auto& name : target_function_names) {
+      auto gvar = mod->GetGlobalVar(name);
+      target_functions.push_back({gvar, Downcast<Function>(mod->Lookup(gvar))});
+    }
+  } else {
+    // Get all the functions that have the `num_input` attribute.
+    for (const auto& [gvar, func] : mod->functions) {
+      if (func->IsInstance<FunctionNode>()) {
+        auto opt_num_input = func->GetAttr<Integer>(attr::kNumInput);
+        if (opt_num_input) {
+          target_functions.emplace_back(gvar, Downcast<Function>(func));
+        }
+      }
+    }
+    std::sort(target_functions.begin(), target_functions.end(),
+              [](const auto& lhs, const auto& rhs) {
+                return lhs.first->name_hint < rhs.first->name_hint;
+              });
+  }
+  return target_functions;
+}
+
 }  // namespace
 
 namespace transform {
 
-Pass PartitionTransformParams() {
+Pass PartitionTransformParams(bool shared_transform, const Array<String>& target_function_names) {
   auto pass_func = [=](IRModule mod, PassContext pc) {
-    PreprocessPartitioner mutator;
-
     IRModule updates;
-    for (const auto& [gvar, func] : mod->functions) {
-      if (auto opt = func.as<relax::Function>()) {
-        auto new_func = Downcast<Function>(mutator(opt.value()));
-        if (!new_func.same_as(func)) {
-          updates->Add(gvar, new_func);
-        }
+
+    std::optional<GlobalCollectInfo> global_collect_info;
+    auto target_functions = GetTargetFunctions(mod, target_function_names);
+
+    if (shared_transform) {
+      std::vector<Function> functions;
+      for (const auto& [_, func] : target_functions) {
+        functions.push_back(func);
       }
+      global_collect_info = MakeGlobalLiftPlan(mod, functions);
+    }
+
+    std::unordered_map<GlobalVar, LocalCollectInfo, ObjectPtrHash, ObjectPtrEqual>
+        local_collect_info;
+    for (const auto& [gvar, func] : target_functions) {
+      auto info = LocalLiftableBindingCollector::Collect(
+          func, global_collect_info.has_value() ? &global_collect_info.value() : nullptr);
+      local_collect_info[gvar] = info;
+    }
+
+    for (const auto& [gvar, info] : local_collect_info) {
+      auto new_runtime_func = info.MakeRuntimeFunction();
+      updates->Add(gvar, new_runtime_func);
+      if (!global_collect_info.has_value()) {
```

Comment:
Can this conditional be moved to an else block of `if (global_collect_info.has_value())`?  That would make it clearer to a reader that the runtime function is always generated, and that there's a choice between a single global transformation and several individual transformations.

```c++
if (global_collect_info.has_value()) {
  auto global_transform = global_collect_info.value().MakeCompileTimeFunc();
  updates->Add(GlobalVar("transform_params"), global_transform);
} else {
  for (const auto& [gvar, info] : local_collect_info) {
      updates->Add(GlobalVar(gvar->name_hint + "_transform_params"),
                     info.MakeCompileTimeFunction());
  }
}
```